### PR TITLE
Default to using Numpy precision in interfaces for arrays

### PR DESCRIPTION
**Pull request by Pinkyboi**
Fixes [#1196](https://github.com/pyccel/pyccel/issues/1196):
Use the default numpy precision when constructing function's arguments in the case of non specified precision for numpy arrays.

### CODE REVIEW

**Comment by EmilyBourne** [comment]
Please add a test that was failing before this PR

**Comment by rmahjoubi** [comment]
can you add a couple more tests such

```python
from pyccel.decorators import types
import numpy as np

@types('int32[:]', 'int32[:]')
@types('float32[:]', 'float32[:]')

def function(a, b):
    print("Hello World!")

if __name__ == '__main__':
    arr = np.array([0, 1, 2, 3, 4], dtype=float)
    function(arr, arr)
```
I think a numeric promotion should be applied to the arguments in the following case, this should be discussed and handled in a different PR , but I think it's a good idea to add a test

```python
from pyccel.decorators import types
import numpy as np

@types('int[:]', 'int[:]')
@types('float[:]', 'float[:]')

def function(a, b):
    print("Hello World!")

if __name__ == '__main__':
    arr = np.array([0, 1, 2, 3, 4], dtype=np.float32)
    function(arr, arr)
```

**Comment by Pinkyboi** [comment]
@rmahjoubi Alright i'll add that kind of test and mark them as expected to fail for now.

**Comment by EmilyBourne** [comment]
> can you add a couple more tests
> 
> I think a numeric promotion should be applied to the arguments in the following case, this should be discussed and handled in a different PR , but I think it's a good idea to add a test

I'm not sure I understand what you mean here. One thing that we have taken a lot of care to avoid is array copies in the wrapper. This is something that f2py does but it's also something that can have a non-negligible effect on performance. That is why the wrapper raises an error explaining the type problem. I don't expect either of your tests to pass ever. Implicit numeric promotion should be discussed for scalars (https://github.com/pyccel/pyccel/discussions/1059), but it should be avoided for arrays.

**Comment by Pinkyboi** [comment]
@nhamidn Any idea what could be causing the failed test when including the `openMP` library in `MacOSX` ?

**Comment by EmilyBourne** [comment]
> @nhamidn Any idea what could be causing the failed test when including the `openMP` library in `MacOSX` ?

I don't see any reason why your PR would cause this. It looks like a change to the setup on mac. If this is the case then other PRs should encounter the same problem. `libomp` is installed on the system, but maybe we need reinstall to ensure it's correctly wired up

**Comment by EmilyBourne** [comment]
> > @nhamidn Any idea what could be causing the failed test when including the `openMP` library in `MacOSX` ?
> 
> I don't see any reason why your PR would cause this. It looks like a change to the setup on mac. If this is the case then other PRs should encounter the same problem. `libomp` is installed on the system, but maybe we need reinstall to ensure it's correctly wired up

It looks like this is a git problem. Libomp has recently been updated from 14.0.6 to 15.0.2

**Comment by rmahjoubi** [comment]
@pinkyboi could you please add similar tests to those I mentioned above, the first one is still failing on this branch

**Comment by EmilyBourne** [comment]
> @Pinkyboi could you please add similar tests to those I mentioned above, the first one is still failing on this branch

@rmahjoubi Which test are you talking about? The first test here : https://github.com/pyccel/pyccel/pull/1202#issuecomment-1284692017 doesn't work on the master branch and would require copying an array

**Comment by rmahjoubi** [comment]
@EmilyBourne I am talking about this test
```python
from pyccel.decorators import types
import numpy as np

@types('float64', 'float64')
@types('int', 'int')

def function(a, b):
    print("Hello World!")

if __name__ == '__main__':
    function(3.4,3.3)
```
It is not related directly to this issue but I think it's a simple fix.
The others I thought we may make sure an exception is raised.

**Comment by EmilyBourne** [comment]
> @EmilyBourne I am talking about this test
> 
> ```python
> from pyccel.decorators import types
> import numpy as np
> 
> @types('float64', 'float64')
> @types('int', 'int')
> 
> def function(a, b):
>     print("Hello World!")
> 
> if __name__ == '__main__':
>     function(3.4,3.3)
> ```
> 
> It is not related directly to this issue but I think it's a simple fix. The others I thought we may make sure an exception is raised.

I am not convinced that this is a simple fix. It is simple to fix it here (if you are familiar with the cwrapper). But this problem should be fixed in all the places it occurs at the same time otherwise it becomes much harder to explain in the documentation. Because of that I would rather leave this to its own PR, after discussion of #1059.

I agree that an error should be raised for the others though. Is that not the case?

**Comment by rmahjoubi** [comment]
> > @EmilyBourne I am talking about this test
> > ```python
> > from pyccel.decorators import types
> > import numpy as np
> > 
> > @types('float64', 'float64')
> > @types('int', 'int')
> > 
> > def function(a, b):
> >     print("Hello World!")
> > 
> > if __name__ == '__main__':
> >     function(3.4,3.3)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It is not related directly to this issue but I think it's a simple fix. The others I thought we may make sure an exception is raised.
> 
> I am not convinced that this is a simple fix. It is simple to fix it here (if you are familiar with the cwrapper). But this problem should be fixed in all the places it occurs at the same time otherwise it becomes much harder to explain in the documentation. Because of that I would rather leave this to its own PR, after discussion of #1059.
> 
> I agree that an error should be raised for the others though. Is that not the case?

I can't see the relation between this test and #1059 as no casting is required there

**Comment by EmilyBourne** [comment]
> I can't see the relation between this test and https://github.com/pyccel/pyccel/discussions/1059 as no casting is required there

In python there would be casting. If we were talking about integers then there would be casting on certain platforms. These are the kind of distinctions that would need to be explained in the docs if we fix this for cases where we know that in C there is no casting

**Comment by rmahjoubi** [comment]
> > I can't see the relation between this test and #1059 as no casting is required there
> 
> In python there would be casting. If we were talking about integers then there would be casting on certain platforms. These are the kind of distinctions that would need to be explained in the docs if we fix this for cases where we know that in C there is no casting

@EmilyBourne Yes now I see

**Review by rmahjoubi — APPROVED** [review]

File: pyccel/ast/core.py [hunk 1]
```diff
@@ -2818,8 +2819,9 @@ def point(self, args, use_final_precision = False):
             type_match = lambda dtype1, dtype2, call_arg, func_arg: \
                     (dtype1 in dtype2 or dtype2 in dtype1) \
                     and (call_arg.rank == func_arg.rank) \
-                    and call_arg.precision == func_arg.precision
-
+                    and (call_arg.precision == (func_arg.precision\
+                        if not func_arg.is_ndarray \
+                        else get_final_precision(func_arg)))
```
[the remark this answers is not in the document]
Comment (Pinkyboi) [reply pyccel/ast/core.py]: @rmahjoubi  It seems like that behaviour is wanted for interfaces to distinguish between `float` and `float64` for example,  or at least that's what i've understood from the comment that @EmilyBourne posted [here](https://github.com/pyccel/pyccel/issues/1196#issuecomment-1267253305)

File: pyccel/ast/core.py [hunk 2]
```diff
@@ -2820,7 +2821,6 @@ def point(self, args, use_final_precision = False):
                     and (call_arg.rank == func_arg.rank) \
                     and call_arg.precision == func_arg.precision
 
-
```
Comment (EmilyBourne) [inline pyccel/ast/core.py:2823]: This file has not changed in a way that is useful for this PR. Please revert the changes to make the PR cleaner. (`git checkout master -- pyccel/ast/core.py`)

File: pyccel/ast/variable.py [hunk 3]
```diff
@@ -123,6 +123,7 @@ def __init__(
         precision=0,
         is_argument=False,
         is_kwonly=False,
+        is_tuple=False,
```
Comment (EmilyBourne) [inline pyccel/ast/variable.py:126]: Why do you need this argument? Shouldn't `is_tuple` be False unless this class is `TupleVariable`?

File: pyccel/ast/variable.py [hunk 4]
```diff
@@ -212,6 +217,10 @@ def __init__(
         if self._rank < 2:
             self._order = None
 
+        if self.is_ndarray and not self.is_tuple\
```
Comment (EmilyBourne) [inline pyccel/ast/variable.py:220]: Are there cases where `self.is_ndarray` and `self.is_tuple`? If so this seems like a bug that needs to be fixed

File: pyccel/ast/variable.py [hunk 5]
```diff
@@ -401,6 +410,12 @@ def is_ndarray(self):
         return isinstance(self.dtype, (NativeInteger, NativeBool,
                           NativeFloat, NativeComplex))
 
+    @property
+    def is_tuple(self):
+        """user friendly method to check if the variable is a tuple:
+        """
+        return self._is_tuple
```
Comment (EmilyBourne) [inline pyccel/ast/variable.py:417]: Shouldn't this return `False`? It should only be `True` if it is overwritten by `TupleVariable.is_tuple`

File: pyccel/parser/semantic.py [hunk 6]
```diff
@@ -503,6 +503,17 @@ def _infere_type(self, expr, **settings):
...
 
+        elif isinstance(expr, IndexedElement):
+            d_var['datatype'   ] = expr.dtype
+            d_var['memory_handling'] = 'heap' if expr.rank > 0 else 'stack'
+            d_var['shape'      ] = expr.shape
+            d_var['rank'       ] = expr.rank
+            d_var['order'      ] = expr.order
+            d_var['precision'  ] = expr.precision
+            d_var['cls_base'   ] = expr.base.cls_base
+            d_var['is_tuple'   ] = expr.base.is_tuple
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:514]: Couldn't this simply be ` d_var['is_tuple'   ] = isinstance(expr.base, TupleVariable)`?

File: pyccel/parser/semantic.py [hunk 7]
```diff
@@ -503,6 +503,17 @@ def _infere_type(self, expr, **settings):
...
+            return d_var
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:515]: I don't think that this is correct. If we have the following code:
```python
a = (1,2,3)
b = a[0]
```
then you will set `b.is_tuple` to True, but `b` is not a tuple

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
I am not entirely convinced by your fix. It seems better to me to create the variable with the correct precision where the problem occurs (arguments) rather than trying to correct the variable. This would reduce the number of times the `if` is encountered and reduces the potential for errors later. For example would it be possible to fix the problem [here](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/headers.py#L287-L288)?

I am not convinced that `is_tuple` is necessary or useful. If you can convince me that it is needed then the docstring of `Variable` needs updating

**Comment by Pinkyboi** [comment]
@EmilyBourne  I totally get your idea, the fix look clanky, even when i was implementing it i had that idea that there must be a better way to do it ( i just didn't know where to modify so i went with parts of the code that i've already touched).

File: tests/epyccel/test_default_precision_template.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    @types('float[:]')
+    @types('complex[:]')
+    def return_array_element(array):
+        return array[0]
+
+    test_types = ['int', 'float', 'complex']
+    f1 = return_array_element
+    f2 = epyccel(f1, language=language)
+    for t in test_types:
+        d1 = randint(1, 15)
```
Comment (aihya) [inline tests/epyccel/test_default_precision_template.py:23]: You sure this test works for all types of complex numbers? I mean `2+3j`, `3j` or `2`?

Comment (aihya) [reply tests/epyccel/test_default_precision_template.py:23]: It seems like it's only using real part of the complex numbers that's why I asked.

Comment (Pinkyboi) [reply tests/epyccel/test_default_precision_template.py:23]: I'm not sure that matters as long as we are testing with a complex type, but i may be wrong.

**Review by aihya — APPROVED** [review]

File: pyccel/ast/headers.py [hunk 9]
```diff
@@ -284,6 +284,10 @@ def build_argument(var_name, dc):
 
             order = None
             shape = None
+
+            if rank and precision == -1:
+                precision = default_precision[dtype]
```
Comment (EmilyBourne) [inline pyccel/ast/headers.py:289]: @aihya @azouiten I just wanted to point out that this syntax may also be used in pyccel to identify arrays, so you will also need to look out for this when handling lists

**Review by EmilyBourne — APPROVED** [review]

**Comment by yguclu** [comment]
@Pinkyboi Thanks for this PR!

I am a bit confused by the title and description. In order to solve issue #1196 you added the lines

```python
            if rank and precision == -1:
                precision = default_precision[dtype]
```

to the local function `build_argument`, which is defined inside method `create_definition` of class `FunctionHeader`.

Here you actually used the **default precision** for Numpy arrays whenever the user does not specify the precision explicitly. So why do you use the words *"Default to using max precision"* and *"Convert default precision type specifiers to max precision"* ?

**Comment by EmilyBourne** [comment]
> @Pinkyboi Thanks for this PR!
> 
> I am a bit confused by the title and description. In order to solve issue #1196 you added the lines
> 
> ```python
>             if rank and precision == -1:
>                 precision = default_precision[dtype]
> ```
> 
> to the local function `build_argument`, which is defined inside method `create_definition` of class `FunctionHeader`.
> 
> Here you actually used the **default precision** for Numpy arrays whenever the user does not specify the precision explicitly. So why do you use the words _"Default to using max precision"_ and _"Convert default precision type specifiers to max precision"_ ?

@yguclu I think what is meant by "default precision" in this PR is `-1`. Which is the precision of native python objects and tells us to use the default value. Perhaps it would be clearer to call this PR "Use numpy types inside numpy arrays"?

**Comment by Pinkyboi** [comment]
@yguclu i've changed the title and description, is it more representative of the actual issue now ?

**Comment by yguclu** [comment]
> @yguclu i've changed the title and description, is it more representative of the actual issue now ?

Yes, it is much clearer now! Thanks

File: tests/epyccel/test_default_precision_template.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        arr = np.ones(d1, dtype=t)
+        python_result = f1(arr)
+        pyccel_result = f2(arr)
+
+        assert isinstance(pyccel_result, type(python_result))
+        assert isclose(pyccel_result, python_result)
```
Comment (yguclu) [inline tests/epyccel/test_default_precision_template.py:29]: In the unit tests we should explicitly define the relative and absolute tolerances for Numpy array comparison, by passing `rtol` and `atol`. See [here](https://github.com/pyccel/pyccel/blob/ca37095162bf364b63abf88f956d66e5bf9bb248/tests/epyccel/test_arrays.py#L3970-L4012) for example.

**Review by yguclu — CHANGES_REQUESTED** [review]

**Review by yguclu — APPROVED** [review]
